sdk/js/src/utils/repairVaa.ts

```diff
@@ -0,0 +1,137 @@
+    result.guardianSet[i] = { keys: guardian_set[0], expiry: guardian_set[1] };
+  }
+  return result;
+}
+
+export async function repairVaa(vaaHex: string, rpcUrl: string) {
+  const provider = new ethers.providers.JsonRpcProvider(rpcUrl);
+  const guardianSetsInfo = await getGuardianSets(provider);
+  const currentGuardianSetIndex = guardianSetsInfo.currentGuardianSetIndex;
+  const guardianSets = guardianSetsInfo?.guardianSet;
+  const guardianSetIndexes = Object.keys(guardianSets);
+  const vaaGuardianSetIndex = parseInt(vaaHex.slice(2, 10));
+  if (vaaGuardianSetIndex === parseInt(currentGuardianSetIndex)) {
+    console.log("Vaa has current guardian set index. no repair needed");
+    return vaaHex;
+  } else if (guardianSetIndexes.includes(vaaGuardianSetIndex.toString())) {
+    console.log(
+      `Vaa is using old GS index=${vaaGuardianSetIndex}. Will attempt to update to current GS Index=${currentGuardianSetIndex}...\n`
+    );
+
+    const currentGuardianSet = guardianSets[currentGuardianSetIndex].keys;
+    const minNumSignatures =
+      Math.floor((2.0 * currentGuardianSet.length) / 3.0) + 1;
+    // console.log(`Current GS: #${currentGuardianSetIndex}`, currentGuardianSet);
+    const numSignatures = parseInt(vaaHex.slice(10, 12), 16);
+    const version = vaaHex.slice(0, 2);
+    console.log(
+      `There are ${numSignatures} signatures found in vaa. There are ${currentGuardianSet.length} Guardians. We need at least ${minNumSignatures}.\n`
+    );
+    try {
+      const parsed_vaa = parseVaa(hexToUint8Array(vaaHex));
+
+      const parsedVaaSignatureLength = parsed_vaa.guardianSignatures.length;
+      if (parsedVaaSignatureLength !== numSignatures) {
+        console.error("# of parsed signatures does not match signature length");
+      }
+      const digest = keccak256(parsed_vaa.hash).toString("hex");
+
+      var validSignatures: GuardianSignature[] = [];
+      // take each signature, check if valid against hash & current guardian set
+      // if valid, keep
+      // if invalid, discard
+      parsed_vaa.guardianSignatures.forEach((signature) => {
+        try {
+          const vaaGuardianPublicKey = ethers.utils.recoverAddress(
+            hex(digest),
+            hex(signature.signature.toString("hex"))
+          );
+
+          const currentIndex = signature.index;
+          const currentGuardianPublicKey = currentGuardianSet[currentIndex];
+
+          if (currentGuardianPublicKey === vaaGuardianPublicKey) {
+            console.log(
+              `found a match for gs index=${currentIndex}, public key=${currentGuardianPublicKey.toString(
+                "hex"
+              )}`
+            );
+            validSignatures.push(signature);
+          } else {
+            console.error(
+              `did not find a match for gs index=${currentIndex}, vaaGuardianPublicKey=${vaaGuardianPublicKey}, current public key=${currentGuardianPublicKey}`
+            );
+          }
+        } catch (err) {
```

Comment:
maybe make the try block narrower? there's too much stuff happening inside to easily follow what exactly we're catching. Also, the exception there looks like a genuine error in the VAA, so might be best to not catch and just bubble it up to the caller - they're in a better position to handle an invalid VAA than this function

Reply:
so it might not be a genuine error in the VAA if that guardian at that index is no longer valid with the guardian set that was provided. and it looks like the recoverAddress function will throw an error if it can't find a valid address... so i kept the try/catch there so it can loop through the rest of the signatures.